src/main/java/com/softserveinc/dokazovi/service/impl/AuthorServiceImpl.java

```diff
@@ -0,0 +1,157 @@
+			DoctorEntity doctor = doctorRepository.save(doctorEntity);
+			dto = toAuthorDTO(doctor.getProfile(), doctor);
+			return dto;
+		}
+		throw new ForbiddenPermissionsException();
+	}
+
+	private AuthorDTO toAuthorDTO(UserEntity savedUser, DoctorEntity savedDoctor) {
+		return AuthorDTO.builder().id(savedDoctor.getId())
+				.socialNetwork(savedDoctor.getSocialNetwork())
+				.avatar(savedUser.getAvatar())
+				.bio(savedDoctor.getBio())
+				.firstName(savedUser.getFirstName())
+				.lastName(savedUser.getLastName())
+				.placeOfWork(savedDoctor.getQualification())
+				.email(savedUser.getEmail())
+				.city(savedDoctor.getCity().getId())
+				.build();
+	}
+
+	private DoctorEntity createDoctorEntity(AuthorDTO authorDTO) {
+		CityEntity city = cityRepository.getOne(authorDTO.getCity());
+		return DoctorEntity.builder()
+				.bio(authorDTO.getBio())
+				.qualification(authorDTO.getPlaceOfWork())
+				.socialNetwork(authorDTO.getSocialNetwork())
+				.publishedPosts(0L)
+				.promotionScale(1.0)
+				.profile(toUserEntity(authorDTO))
+				.promotionLevel(UserPromotionLevel.BASIC)
+				.city(city)
+				.build();
+	}
+
+	private UserEntity toUserEntity(AuthorDTO authorDTO) {
+		return UserEntity.builder()
+				.id(null)
+				.firstName(authorDTO.getFirstName())
+				.lastName(authorDTO.getLastName())
+				.email(authorDTO.getEmail())
+				.status(UserStatus.ACTIVE)
+				.enabled(true)
+				.role(RoleEntity.builder()
+						.id(3)
+						.name("Doctor")
+						.permissions(Set.of(RolePermission.SAVE_OWN_PUBLICATION))
+						.build())
+				.password(passwordEncoder.encode("Kolala"))
+				.createdAt(Timestamp.valueOf(LocalDateTime.now()))
+				.build();
+	}
+
+	@Override
+	public AuthorDTO update(AuthorDTO authorDTO, UserPrincipal userPrincipal) {
+		if (userPrincipal.getAuthorities().stream()
+				.anyMatch(
+						grantedAuthority -> grantedAuthority.getAuthority()
+								.equals(RolePermission.UPDATE_AUTHOR.getAuthority()))) {
+			DoctorEntity doctor = doctorRepository.getOne(authorDTO.getId());
+			doctor.setBio(authorDTO.getBio());
+			doctor.setQualification(authorDTO.getPlaceOfWork());
+			doctor.setCity(cityRepository.getOne(authorDTO.getCity()));
+			doctor.setSocialNetwork(authorDTO.getSocialNetwork());
+			UserEntity user = userRepository.getOne(doctor.getProfile().getId());
+			user.setEmail(authorDTO.getEmail());
+			user.setFirstName(authorDTO.getFirstName());
+			user.setLastName(authorDTO.getLastName());
+			user.setDoctor(doctor);
+			doctor.setProfile(user);
+			DoctorEntity savedDoctor = doctorRepository.save(doctor);
+			return toAuthorDTO(savedDoctor.getProfile(), savedDoctor);
+		}
+		throw new ForbiddenPermissionsException();
+	}
+
+	@Override
+	public Boolean delete(Integer authorId, UserPrincipal userPrincipal) {
+		if (userPrincipal.getAuthorities().stream()
+				.anyMatch(
+						grantedAuthority -> grantedAuthority.getAuthority()
+								.equals(RolePermission.DELETE_AUTHOR.getAuthority()))) {
```

Comment:
I'd suggest creating a universal private method `hasAuthority(UserPrincipal userPrincipal, RolePermission permision)`  that contains the following logic `userPrincipal.getAuthorities().stream()
				.anyMatch(
						grantedAuthority -> grantedAuthority.getAuthority()
								.equals(RolePermission.DELETE_AUTHOR.getAuthority()))`

Reply:
fixed